### PR TITLE
Fix for issue #407: Prevent crashes due to unbounded growth.

### LINKED ISSUE #407: Crash by repeatedly doubling list size and running out of memory
Running on:
- minecraft 1.19.2
- quilt loader 0.17.10
- hex casting 0.10.3

I removed other mods to make sure it was not something else causing this.

image

[crashreport.txt](https://github.com/gamma-delta/HexMod/files/10608367/crashreport.txt)

### PR DESCRIPTION
**Pull request by navarchus**
Prevent crashes due to unbounded growth of a few items.

Fixes #407.

### CODE REVIEW

**Comment by navarchus** [comment]
Special thanks to @walksanatora for helping me learn the repo

**Comment by vgskye** [comment]
shouldn't this use the limit at https://github.com/FallingColors/HexMod/blob/main/Common/src/main/java/at/petrak/hexcasting/common/lib/hex/HexIotaTypes.java#L26?

**Comment by navarchus** [comment]
> shouldn't this use the limit at https://github.com/FallingColors/HexMod/blob/main/Common/src/main/java/at/petrak/hexcasting/common/lib/hex/HexIotaTypes.java#L26?

I believe this is a different enough scenario, as the bug in question is about a spell which continually adds to a list whereas these constants seem to be more related to trying to limit the max size of a spell before it is cast.

**Comment by vgskye** [comment]
> > shouldn't this use the limit at [`main`/Common/src/main/java/at/petrak/hexcasting/common/lib/hex/HexIotaTypes.java#L26](https://github.com/FallingColors/HexMod/blob/main/Common/src/main/java/at/petrak/hexcasting/common/lib/hex/HexIotaTypes.java?rgh-link-date=2024-08-20T03%3A15%3A22Z#L26)?
> 
> I believe this is a different enough scenario, as the bug in question is about a spell which continually adds to a list whereas these constants seem to be more related to trying to limit the max size of a spell before it is cast.

I believe the MAX_SERIALIZATION_TOTAL limit is commonly known as serialization limit, which.... limits hexes in much the same way as this pull in practice, so I feel like treating this limit as an extension of that is reasonable.
(also it'd let you reuse IotaType.isTooLargeToSerialize for checking)

File: Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/CastingVM.kt [hunk 1]
```diff
@@ -50,15 +51,30 @@ class CastingVM(var image: CastingImage, val env: CastingEnvironment) {
             // ...and execute it.
             // TODO there used to be error checking code here; I'm pretty sure any and all mishaps should already
             // get caught and folded into CastResult by evaluate.
-            val image2 = next.evaluate(continuation.next, world, this)
+            val image2 = next.evaluate(continuation.next, world, this).let { result ->
+                // if stack is unable to be serialized, have the result be an error
+                if (result.newData != null && isTooLargeToSerialize(result.newData.stack)) {
```
Comment (object-Object) [inline Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/CastingVM.kt]: Nit: this should be `IotaType.isTooLargeToSerialize` imo - I don't think importing the static method is really necessary.

File: Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/CastingVM.kt [hunk 2]
```diff
@@ -50,15 +51,30 @@ class CastingVM(var image: CastingImage, val env: CastingEnvironment) {
...
+                } else {
+                    result
+                }
+            }
+
             // Then write all pertinent data back to the harness for the next iteration.
             if (image2.newData != null) {
                 this.image = image2.newData
             }
+
```
Comment (object-Object) [inline Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/CastingVM.kt]: Unnecessary whitespace change.

File: Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/CastingVM.kt [hunk 3]
```diff
@@ -50,15 +51,30 @@ class CastingVM(var image: CastingImage, val env: CastingEnvironment) {
...
             this.env.postExecution(image2)
 
             continuation = image2.continuation
             lastResolutionType = image2.resolutionType
+
```
Comment (object-Object) [inline Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/CastingVM.kt]: Unnecessary whitespace change.

File: Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/CastingVM.kt [hunk 4]
```diff
@@ -50,15 +51,30 @@ class CastingVM(var image: CastingImage, val env: CastingEnvironment) {
             // ...and execute it.
             // TODO there used to be error checking code here; I'm pretty sure any and all mishaps should already
             // get caught and folded into CastResult by evaluate.
-            val image2 = next.evaluate(continuation.next, world, this)
+            val image2 = next.evaluate(continuation.next, world, this).let { result ->
+                // if stack is unable to be serialized, have the result be an error
+                if (result.newData != null && isTooLargeToSerialize(result.newData.stack)) {
+                    result.copy(
+                        newData = null,
+                        sideEffects = result.sideEffects + OperatorSideEffect.DoMishap(MishapStackSize(), Mishap.Context(null, null)),
```
Comment (object-Object) [inline Common/src/main/java/at/petrak/hexcasting/api/casting/eval/vm/CastingVM.kt]:
```suggestion
                        sideEffects = listOf(OperatorSideEffect.DoMishap(MishapStackSize(), Mishap.Context(null, null))),
```

File: Common/src/main/resources/assets/hexcasting/lang/en_us.flatten.json5 [hunk 5]
```diff
@@ -901,6 +901,7 @@
       invalid_spell_datum_type: "Tried to use a value of invalid type as a SpellDatum: %s (class %s). This is a bug in the mod.",
       unknown: "threw an exception (%s). This is a bug in the mod.",
       shame: "Shame on you!",
+      stack_size: "My mind was too small to contain the breadth of this spell.",
```
Comment (object-Object) [inline Common/src/main/resources/assets/hexcasting/lang/en_us.flatten.json5]: I'd argue this should be a bit less flowery, and actually say what the player did wrong (ie. exceed the stack size limit), to match other error messages.

File: Forge/src/main/java/at/petrak/hexcasting/forge/ForgeHexConfig.java
Comment (object-Object) [inline Forge/src/main/java/at/petrak/hexcasting/forge/ForgeHexConfig.java:1]: Unnecessary whitespace changes.

**Review by object-Object — CHANGES_REQUESTED** [review]

File: Common/src/main/resources/assets/hexcasting/lang/en_us.flatten.json5 (hunk 5 above)
Comment (navarchus) [reply Common/src/main/resources/assets/hexcasting/lang/en_us.flatten.json5]: made more literal

File: Common/src/main/resources/assets/hexcasting/lang/en_us.flatten.json5 [hunk 6]
```diff
@@ -901,6 +901,7 @@
       invalid_spell_datum_type: "Tried to use a value of invalid type as a SpellDatum: %s (class %s). This is a bug in the mod.",
       unknown: "threw an exception (%s). This is a bug in the mod.",
       shame: "Shame on you!",
+      stack_size: "Exceeded stack size limit",
```
Comment (object-Object) [inline Common/src/main/resources/assets/hexcasting/lang/en_us.flatten.json5]:
```suggestion
      stack_size: "Exceeded the size limit of the stack",
```

**Review by object-Object — APPROVED** [review]
Looks good. It would be nice if the error message showed the pattern that caused it, but I'll do that myself later. Thanks for the PR!

**Pull request merged**